homeassistant/components/stream/hls.py

```diff
@@ -82,29 +109,151 @@ class HlsPlaylistView(StreamView):
+
+        This method is run on every playlist request and the output is not saved.
+        """
+        # Create a copy of the segment with the parts frozen so we can iterate safely
+        segment_copy = copy(segment)
+        segment_copy.parts_by_http_range = segment_copy.parts_by_http_range.copy()
+        (
+            segment.hls_playlist,
+            segment.hls_playlist_parts,
+            segment.hls_num_parts_rendered,
+            segment.hls_playlist_complete,
+        ) = cls.render_segment(segment_copy, ll_hls, last_stream_id)
+        rendered = segment.hls_playlist.format(segment.hls_playlist_parts)
+        playlist = [rendered] if rendered else []
+
+        # Add preload hint
+        # pylint: disable=undefined-loop-variable
+        if ll_hls:
+            if segment_copy.complete:  # Next part belongs to next segment
+                playlist.append(
+                    f'#EXT-X-PRELOAD-HINT:TYPE=PART,URI="./segment/{segment_copy.sequence+1}'
+                    '.m4s",BYTERANGE-START=0'
+                )
+            else:  # Next part is in the same segment
+                if segment_copy.parts_by_http_range:
+                    last_part = next(
+                        reversed(segment_copy.parts_by_http_range.values())
+                    )
+                    start = last_part.http_range_start + len(last_part.data)
+                else:
+                    start = 0
+                playlist.append(
+                    f'#EXT-X-PRELOAD-HINT:TYPE=PART,URI="./segment/{segment_copy.sequence}'
+                    f'.m4s",BYTERANGE-START={start}'
+                )
+        return "\n".join(playlist)
+
     @staticmethod
-    def render(track: StreamOutput) -> str:
-        """Render playlist."""
+    def render_segment(
+        segment: Segment, ll_hls: bool, last_stream_id: int
+    ) -> tuple[str, str, int, bool]:
+        """Render a segment.
+
+        Return a base string, a ll_hls string that can be interjected, the number of parts
+        rendered so far, and a bool that indicates if the rendering for the segment is complete.
+        """
+        playlist = [segment.hls_playlist] if segment.hls_playlist else []
+        playlist_parts = (
+            [segment.hls_playlist_parts] if segment.hls_playlist_parts else []
+        )
+        if not playlist:
+            if last_stream_id != segment.stream_id:
+                playlist = [
+                    "#EXT-X-DISCONTINUITY",
+                    "#EXT-X-PROGRAM-DATE-TIME:"
+                    + segment.start_time.strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3]
+                    + "Z",
+                ]
+            playlist.append("{}")
+        if ll_hls:
+            for part in itertools.islice(
+                segment.parts_by_http_range.values(),
+                segment.hls_num_parts_rendered,
+                None,
+            ):
+                playlist_parts.append(
+                    f"#EXT-X-PART:DURATION={part.duration:.3f},URI="
```

Comment:
Is it helpful to use partial segments even in the non-hls mode? I was reading the spec and saw this appear even without considering low latency, but i don't know if there are implications for doing that.

Reply:
I don't think they are useful outside of LL-HLS

Reply:
Are they ignored or harmful? I was assuming that clients that don't support LL HLS could ignore them, and the code could get simpler. Possibly a bad idea, so feel free to disregard if this is a problem.

Reply:
I think they are ignored. I think putting this behind an `if` was just to try and avoid extra overhead in the non LL-HLS case, but if it's confusing we can remove it.